File: Data.Resumption/Execution/ServiceContext.cs
using System;
using System.Collections.Generic;
using Data.Resumption.Services;

namespace Data.Resumption.Execution
{
    internal class ServiceContext : IServiceContext, IDisposable
    {
        private class Cache : IDisposable
        {
            private readonly Dictionary<Type, object> _services = new Dictionary<Type, object>();
            private readonly Stack<IDisposable> _disposalStack = new Stack<IDisposable>();

            public bool TryGetService(Type ty, out object service) => _services.TryGetValue(ty, out service);
            public void CacheService(Type ty, object service)
            {
                _services[ty] = service;
                var disposable = service as IDisposable;
                if (disposable != null) _disposalStack.Push(disposable);
            }

            public void Dispose()
            {
                var exceptions = new List<Exception>();
                while (_disposalStack.Count > 0)
                {
                    try
                    {
                        _disposalStack.Pop().Dispose();
                    }
                    catch (Exception ex)
                    {
                        exceptions.Add(ex);
                    }
                }
                if (exceptions.Count == 1) throw exceptions[0];
                if (exceptions.Count > 1) throw new AggregateException(exceptions);
            }
        }
        private Cache _execution = new Cache();
        private Cache _step;

        private readonly IServiceFactory _factory;
        private readonly object _sync = new object();

        public ServiceContext(IServiceFactory factory)
        {
            _factory = factory;
        }

        public TService GetService<TService>()
        {
            lock (_sync)
            {
                var ty = typeof(TService);
                object service;
                if (_step != null && _step.TryGetService(ty, out service)
                    || _execution.TryGetService(ty, out service))
                {
                    return (TService)service;
                }
                var living = _factory.CreateService<TService>(this);
                if (living == null) throw new NotSupportedException($"The service type {ty} is not supported by the service factory");
                switch (living.Value.Lifetime)
                {
                    case ServiceLifetime.ExecutionContext:
                        _execution.CacheService(ty, living.Value.Service);
                        break;
                    case ServiceLifetime.ExecutionStep:
                        _step.CacheService(ty, living.Value.Service);
                        break;
                    default:
                        throw new ArgumentOutOfRangeException(nameof(living.Value.Lifetime));
                }
                return living.Value.Service;
            }
        }

        public void BeginStep()
        {
            lock (_sync)
            {
                _step?.Dispose();
                _step = new Cache();
            }
        }

        public void EndStep()
        {
            lock (_sync)
            {
                _step?.Dispose();
                _step = null;
            }
        }

        public void Dispose()
        {
            lock (_sync)
            {
                EndStep();
                _execution?.Dispose();
                _execution = null;
            }
        }
    }
}

File: Data.Resumption.Workflows/ServiceContext.fs
namespace Data.Resumption
open System
open System.Collections.Generic
    

[<AllowNullLiteral>]
type private ServiceContextCache() =
    let services = new Dictionary<Type, obj>()
    let disposalStack = new Stack<IDisposable>()

    member __.TryGetService(ty, svc : obj byref) =
        services.TryGetValue(ty, &svc)
    member __.CacheService(ty, svc : obj) =
        services.[ty] <- svc
        match svc with
        | :? IDisposable as disp ->
            disposalStack.Push(disp)
        | _ -> ()
    member __.Dispose() =
        let exns = new ResizeArray<exn>()
        while disposalStack.Count > 0 do
            try
                disposalStack.Pop().Dispose()
            with
            | exn -> exns.Add(exn)
        if exns.Count > 0 then
            raise (aggregate exns)
    interface IDisposable with
        member this.Dispose() = this.Dispose()


type ServiceContext(factory : ServiceFactory) =
    let factory = new DefaultServiceFactory(factory)
    let execution = new ServiceContextCache()
    let sync = new obj()
    let mutable step : ServiceContextCache = null

    member __.BeginStep() =
        lock sync <| fun () ->
            if not (isNull step) then step.Dispose()
            step <- new ServiceContextCache()
    member __.EndStep() =
        lock sync <| fun () ->
            if not (isNull step) then step.Dispose()
            step <- null

    member public this.Dispose() =
        lock sync <| fun () ->
            this.EndStep()
            execution.Dispose()

    member public __.GetService<'svc>() : 'svc =
        lock sync <| fun () ->
            let ty = typeof<'svc>
            let mutable cached : obj = null
            if (not (isNull step) && step.TryGetService(ty, &cached)
                || execution.TryGetService(ty, &cached)) then Unchecked.unbox cached else
            let living = factory.CreateService<'svc>()
            if isNull living then
                notSupported (sprintf "The service type %O is not supported by the service factory" ty)
            else
            match living.Lifetime with
            | ServiceLifetime.ExecutionLocal ->
                execution.CacheService(ty, living.Service)
            | ServiceLifetime.StepLocal ->
                if isNull step then logicFault "Can't get step-local service outside of a step"
                step.CacheService(ty, living.Service)
            | _ -> invalidArg "lifetime" (sprintf "Unknown lifetime %d" (int living.Lifetime))
            living.Service

    interface IDisposable with
        member this.Dispose() = this.Dispose()
    